aspnet/fundamentals/dependency-injection.rst

```diff
@@ -1,10 +1,277 @@
+
+.. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Controllers/CharactersController.cs
+	:language: c#
+	:linenos:
+	:lines: 1-32
+	:emphasize-lines: 10,12,14,20,23-27
+
+The `ICharacterRepository` simply defines the two methods the controller needs to work with `Character` instances.
+
+.. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Interfaces/ICharacterRepository.cs
+	:language: c#
+	:linenos:
+
+This interface is in turn implemented by a concrete type, ``CharacterRepository``, that is used at runtime.
+
+.. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Models/CharacterRepository.cs
+	:language: c#
+	:linenos:
+
+Note that ``CharacterRepository`` requests an ``ApplicationDbContext`` in its constructor. It is not unusual for dependency injection to be used in a chained fashion like this, with each requested dependency in turn requesting its own dependencies. The container is responsible for resolving all of the dependencies in the tree and returning the fully resolved object graph.
+
+In this case, both ``ICharacterRepository`` and in turn ``ApplicationDbContext`` must be registered with the services container in ``ConfigureServices`` in ``Startup``. ``ApplicationDbContext`` is configured via the call to the extension method ``AddEntityFramework`` which includes an extension for adding a ``DbContext`` (``AddDbContext<T>``). Registration of the repository is done at the bottom end of ``ConfigureServices``:
+
+.. literalinclude:: dependency-injection/sample/src/DependencyInjectionSample/Startup.cs
+	:language: c#
+	:lines: 48-54, 82-86
+	:linenos:
+	:dedent: 8
+	:emphasize-lines: 6, 11
+
+Entity Framework contexts should be added to the services container using the ``Scoped`` lifetime. This is taken care of automatically if you use the helper methods as shown above. Repositories that will make use of Entity Framework should use the same lifetime.
+
+Service Lifetimes and Registration Options
+------------------------------------------
+
+ASP.NET services can be configured with the following lifetimes:
+
+Transient
+	Transient lifetime services are created each time they are requested. This lifetime works best for lightweight, stateless service.
+
+Scoped
+	Scoped lifetime services are created once per request.
+
+Singleton
+	Singleton lifetime services are created the first time they are requested, and then every subsequent request will use the same instance. If your application requires singleton behavior, this is recommended in place of implementing the singleton design pattern in your class yourself.
+
+Instance
+	You can choose to add an instance directly to the services container. If you do so, this instance will be used for all subsequent requests (this technique will create a Singleton-scoped instance).
```

Comment:
> this technique will create a Singleton-scoped instance

Might confuse readers. `Singleton` services are lazy-loaded while `Instance` services are created right in the `ConfigureServices()` method.